lighthouse-core/audits/audit.js

```diff
@@ -81,6 +81,26 @@ class Audit {
     return clampTo2Decimals(percentile);
   }
 
+  /**
+   * @param {LH.Audit.Details.Table['headings']|LH.Audit.Details.Opportunity['headings']} headings
+   * @param {LH.Audit.Details.Opportunity['items']|LH.Audit.Details.Table['items']} items
+   */
+  static assertHeadingKeysExist(headings, items) {
+    // If there are no items, there's nothing to check.
+    if (!items.length) return;
+    // Only do this in tests for now.
+    if (typeof describe === 'undefined' && !process.env.CI) return;
```

Comment:
I don't really like this. I'd prefer to do something like `process.env.LIGHTHOUSE_UNDER_TEST` like I've done in other projects if others are open to it?

Reply:
> I don't really like this. I'd prefer to do something like `process.env.LIGHTHOUSE_UNDER_TEST` like I've done in other projects if others are open to it?

I'm not wild about setting a precedent and seeing that spread, though :) And it's kind of a bummer to only catch it in CI and not through local testing (I guess `yarn unit:core` could set the flag but `yarn jest my-audit` wouldn't).

I think it's probably good without the CI check, actually. It might be different if we were constructing table columns dynamically, but AFAIK all headers are hardcoded in their audits (and our general design approach and i18n means we'll likely continue doing that), so if a LH dev runs their audit and gets any kind of table output from it (unit, smoke, or sample_v2) it should hit this and get corrected before it could ever throw for an end user.

Reply:
> I'm a little hesitant for removing the entire `if` if we currently have cases where rows are conditionally one type or another and a key might be validly missing in some situations. I don't _think_ that's the case, but with 100+ audits I'm not sure and the failure mode is tanking the entire audit's results.

Yeah, I didn't think about the fact that a row value can be set to undefined to have an empty cell, and it's perfectly possible for a table to end up with rows that all happen to have an empty cell there. That would happen with `performance-budgets`, for instance, if either the number of requests or size went over budget but not both, but it would pass this test because they're explicitly set to `undefined`.

Since in theory this assertion will force audit authors to set explicit `undefined` to avoid false negatives anyways (if it's possible to have a table like that you should probably have a test for it, at which point you have to set an explicit `undefined` so this check passes in CI) it would be better to just be a requirement, rather than a requirement for any code paths that happen to run in tests.

If the blocker is going through the existing audits with tables, that's not insurmountable. We could divide them up pretty easily (I see 40 `makeTableDetails` and 5 `makeOpportunityDetails` calls)